apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/CacheHolder.kt

```diff
@@ -0,0 +1,27 @@
+package com.apollographql.apollo3.cache.normalized.internal
+
+import com.apollographql.apollo3.cache.normalized.ReadOnlyNormalizedCache
+
+class CacheHolder(producer: () -> OptimisticCache) {
```

Comment:
I'm not sure what the purpose of the producer here as when the `CacheHolder ` is created prodicer will be called right away to initialize `optimisticCache ` property.

Also overall question, not sure what is the purpose of this class? As it looks like a wrapped around `ReentrantReadWriteLock`?

Reply:
I'm planning to turn this into an interface. We would have:

```kotlin
interface CacheHolder {
  fun <T> read(block: (ReadOnlyNormalizedCache) -> T)

  fun <T> write(block: (OptimisticCache) -> T): T

  fun writeAndForget(block: (OptimisticCache) -> Unit) 
}
```

And two different implementations:

```kotlin
# Uses Reentrant locks on the JVM
class LockCacheHolder(producer: () -> OpimisticCache): CacheHolder(producer)
# Uses a singleThreadExecutor on the JVM and a serial Queue on Apple
expect class IsolateCacheHolder(producer: () -> OpimisticCache): CacheHolder(producer)
```

You're right that the constructor doesn't belong to the interface and doesn't need to be a producer. I'll change to.

```kotlin
class LockCacheHolder(optimisticCache: OpimisticCache): CacheHolder
```

The long term plan is to have an option during `ApolloStore creation`:

```kotlin
enum JvmConcurencyMode {
  // Uses a ReadWriteLock (SQLite cannot do concurrent reads but memory cache can)
  LOCK,
  // Uses an Isolate pattern, like on native. Incurs some thread switching costs but alllows to do the writes asyncs
  ISOLATE
}
val store = ApolloStore(cacheFactory: NormalizedCacheFactory, cacheKeyResolver: CacheKeyResolver, jvmConcurrencyMode: JvmConcurencyMode)
```